lib/services/expiry/expiry.go

```diff
@@ -0,0 +1,248 @@
+func New(ctx context.Context, cfg *Config) (*Service, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	s := &Service{
+		Config: cfg,
+		ctx:    ctx,
+	}
+	return s, nil
+}
+
+// Start starts the expiry service.
+func (s *Service) Start() error {
+
+	semCfg := services.SemaphoreLockConfigWithRetry{
+		SemaphoreLockConfig: services.SemaphoreLockConfig{
+			Service: s.AccessPoint,
+			Params: types.AcquireSemaphoreRequest{
+				SemaphoreKind: types.KindAccessRequest,
+				SemaphoreName: semaphoreName,
+				MaxLeases:     1,
+				Holder:        s.HostID,
+			},
+			Expiry: semaphoreExpiration,
+			Clock:  s.Clock,
+		},
+		Retry: retryutils.LinearConfig{
+			Clock:  s.Clock,
+			First:  time.Second,
+			Step:   semaphoreExpiration / 2,
+			Max:    semaphoreExpiration,
+			Jitter: retryutils.DefaultJitter,
+		},
+	}
+
+	poll := interval.New(interval.Config{
+		Duration:      scanInterval,
+		FirstDuration: retryutils.FullJitter(scanInterval),
+		Jitter:        retryutils.SeventhJitter,
+		Clock:         s.Clock,
+	})
+	defer poll.Stop()
+
+	for {
+		lease, err := services.AcquireSemaphoreLockWithRetry(
+			s.ctx,
+			semCfg,
+		)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		s.processRequests()
+
+		ctx, cancel := context.WithCancel(lease)
+		defer cancel()
+		lease.Stop()
+		if err := lease.Wait(); err != nil {
+			s.Log.WarnContext(ctx, "error cleaning up semaphore", "error", err)
+		}
+		select {
+		case <-s.ctx.Done():
+			return nil
+		case <-poll.Next():
+		}
+	}
+}
+func (s *Service) processRequests() error {
+	requestsExpired := 0
+	nextPageStart := ""
+	for {
+		var page []*types.AccessRequestV3
+		var err error
+		page, nextPageStart, err = s.getNextPageOfAccessRequests(nextPageStart)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		if len(page) == 0 || requestsExpired >= maxExpiresPerCycle {
+			return nil
+		}
+		minPageDelay := time.After(retryutils.SeventhJitter(minPageDelay))
+		for _, req := range page {
+			if !s.shouldExpire(req) {
+				continue
+			}
+			if err := s.expireRequest(s.ctx, req); err != nil {
+				return trace.Wrap(err)
+			}
+			requestsExpired++
+			if requestsExpired >= maxExpiresPerCycle {
+				return nil
+			}
+		}
+		<-minPageDelay
+	}
+}
+
+func (s *Service) getNextPageOfAccessRequests(startKey string) ([]*types.AccessRequestV3, string, error) {
+	req := &proto.ListAccessRequestsRequest{
+		Sort:       proto.AccessRequestSort_CREATED,
+		Descending: true,
+		Limit:      accessRequestPageLimit,
+		StartKey:   startKey,
+	}
+	resp, err := s.AccessPoint.ListAccessRequests(s.ctx, req)
+	if err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+
+	return resp.AccessRequests, resp.NextKey, nil
+}
+
+func (s *Service) shouldExpire(req types.AccessRequest) bool {
+	expires := req.Expiry()
+	if req.GetState() == types.RequestState_PENDING {
+		expires = expires.Add(pendingRequestGracePeriod)
+	}
+	return s.Clock.Now().After(expires)
```

Comment:
nit: its a subtle distinction, but we really aught to only be expiring values that were past their expiration time *at the time of reading*.  Given that, I'd suggest taking a single value for `now` before loading the current page of requests, then passing it into this function.